index/segment/mem/segment_simple.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+
+	"github.com/uber-go/atomic"
+)
+
+// TODO(prateek): investigate impact of native heap
+type simpleSegment struct {
+	opts     Options
+	id       segment.ID
+	docIDGen *atomic.Uint32
+
+	// internal docID -> document
+	documentsLock sync.RWMutex
+	documents     map[segment.DocID]document // TODO(prateek): measure perf impact of slice v map here
```

Comment:
When I find myself prefixing fields the same I find it best to make it composite.

For instance.

Reply:
Sure thing.

Reply:
ended up using anonymous structs for this:

```
type simpleSegment struct {
  // ...
  docs struct {
    sync.RWMutex
    values map[segment.DocID]document
  }
}
```